### PR TITLE
[spirv] Support Vulkan 1.2 - remove BufferBlock

### PR DESCRIPTION
**Pull request by jaebaek**
We cannot use BufferBlock decoration for SPIR-V 1.4 or above.
Instead, we must use Block decorated StorageBuffer Storage Class.

### CODE REVIEW

File: tools/clang/lib/SPIRV/SpirvContext.cpp [hunk 1]
```diff
@@ -221,6 +221,12 @@ const SpirvPointerType *SpirvContext::getPointerType(const SpirvType *pointee,
       return foundSC->second;
   }
 
+  if (spvTargetEnv >= SPV_ENV_VULKAN_1_2) {
```
Comment (ehsannas) [inline tools/clang/lib/SPIRV/SpirvContext.cpp]: this is actually a bug. You are checking the type cache using the old storage class, and the inserting into the cache using the new storage class.

**Comment by jaebaek** [comment]
Thank you for the comment.

I quickly checked your first approach. I found one corner case and want to discuss it with you.

```
struct S { int4 a; float b; }
RWStructuredBuffer<S> values;
```

the above HLSL code will generate the following SPIR-V code:

```
%type_RWStructuredBuffer_S = OpTypeStruct %_runtimearr_S
%_ptr_StorageBuffer_type_RWStructuredBuffer_S = OpTypePointer StorageBuffer %type_RWStructuredBuffer_S
%values = OpVariable %_ptr_StorageBuffer_type_RWStructuredBuffer_S StorageBuffer
```

Let's assume that we have some SPIR-V code that access `int4 a` of `struct S` using `OpAccessChain`.
It generates `%43 = OpAccessChain %_ptr_Uniform_v4int %values ...`.
See its type `%_ptr_Uniform_v4int` which is not `StorageBuffer`.

*Even though we change the type of pointer or variable in EmitVisitor or a pass just before EmitVisitor, we cannot change all pointer types used by instructions*.

I think we must do it before `LowerTypeVisitor`.

What do you think?

**Comment by ehsannas** [comment]
Yeah, sorry I forgot about OpAccessChain. I still think it can be best done using Method#2 mentioned above. I still feel that the current implementation (e.g. adding the vulkan target environment to the spirv context) is a bit hacky. Let me think about it a bit more in details and let you know.

**Comment by jaebaek** [comment]
Yeah, I agree that we better add a new visitor, but I think it should be invoked before `LowerTypeVisitor`. We might have instructions which must specify pointer types (not only `OpAccesschain`) .. but I am not 100% sure. As far as I remember, some image instructions use pointer types.

**Comment by jaebaek** [comment]
Since @ehsannas did refactoring for this (see #2790 ), I close this PR.

**Pull request closed without merging**